source/usb/webusb/usbd_core_webusb.c

```diff
@@ -0,0 +1,94 @@
+/**
+ * @file    usbd_core_webusb.c
+ * @brief   WebUSB Device driver
+ *
+ * DAPLink Interface Firmware
+ * Copyright (c) 2009-2016, ARM Limited, All Rights Reserved
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "string.h"
+
+#include "RTL.h"
+#include "rl_usb.h"
+#include "usb_for_lib.h"
+#include "info.h"
+
+/*
+ *  USB Device Endpoint 0 Event Callback - WebUSB specific handling (Setup Request To Device)
+ *    Parameters:      none
+ *    Return Value:    TRUE - Setup vendor request ok, FALSE - Setup vendor request not supported
+ */
+
+__weak BOOL USBD_EndPoint0_Setup_WebUSB_ReqToDevice(void)
+{
+    U8  *pD;
+    U32 len, n;
+
+    BOOL success = (__FALSE);
+    if (USBD_SetupPacket.bRequest == usbd_webusb_vendor_code) {			/* vendor code correct? */
+        switch (USBD_SetupPacket.wIndex) {
+            case WEBUSB_REQUEST_GET_ALLOWED_ORIGINS:
+                pD = (U8 *)USBD_WebUSBAllowedOriginsHeader;
+                USBD_EP0Data.pData = pD;
+                len = ((WEBUSB_ALLOWED_ORIGINS_HEADER_DESCRIPTOR *)pD)->wTotalLength;
+                success = (__TRUE);
+                break;
+
+            case WEBUSB_REQUEST_GET_URL:
+                pD = (U8 *)USBD_WebUSBURLDescriptor;
+                if (USBD_SetupPacket.wValueL == 0) {
+                    success = (__FALSE);
+                    break;
+                }
+
+                for (n = 0; n + 1 < USBD_SetupPacket.wValueL; n++) {
+                    if (((WEBUSB_URL_DESCRIPTOR *)pD)->bLength != 0) {
+                        pD += ((WEBUSB_URL_DESCRIPTOR *)pD)->bLength;
+                    }
+                }
+
+                if (((WEBUSB_URL_DESCRIPTOR *)pD)->bLength == 0) {
+                    success = (__FALSE);
+                    break;
+                }
+                strcat(((WEBUSB_URL_DESCRIPTOR *)pD)->URL, "&bid=");
```

Comment:
At the moment I've added only board id as parameter in landing page. It's worth considering adding also additional parameters. For example DAPLink version.

Reply:
Since you're including the board id in the query string here, it's probably no longer necessary to include the USB VID - I assume the board id is enough to derive the vendor id for filtering purposes anyways.

For filtering, the best approach would be to pass the serial number in the query string, which the landing page can use to ensure that the USB device picker shows only the device that initiated the landing page notification. Plus the serial number encodes the target id, HIC id, board id, and version anyways.

Since WebUSB is only permitted over HTTPS, I think this should be safe against exposing the user to fingerprinting as long as:

- All outgoing links have the `rel="noreferrer"` attribute set so that they don't leak the serial number via the referrer header
- No untrusted third-party advertising/analytics javascript is included that might log the URL/query string.

Reply:
Thanks, I like the idea of extending functionality to send full serial number. I'll make this change.

Reply:
We discussed it and decided to remove landing page functionality for now. I'm going to remove it.